File: android-companion/app/src/main/java/com/oneinfinity/companion/frida/FridaManager.kt
package com.oneinfinity.companion.frida

import android.util.Log
import org.json.JSONObject
import java.io.File
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * FridaManager — frida-server lifecycle on the companion device.
 *
 * ARCHITECTURE NOTE (Fix #11):
 * The actual Frida script execution (`frida -U -f <pkg> -l <script>`) runs
 * on the HOST machine that has adb access, NOT inside this Android app.
 * The companion app's role is:
 *   1. Manage frida-server binary on the device (/data/local/tmp/frida-server)
 *   2. Receive {type:frida_inject} commands from the backend
 *   3. Acknowledge receipt — the BACKEND runs "frida -U" on the host side
 *   4. Stream {type:frida_output} results back from the backend to the app
 *
 * Fix #9: isFridaServerAvailable() uses `ps` via `su` instead of `frida-ps`
 *         (which is a host binary and doesn't exist on the Android device).
 * Fix #10: startFridaServer() and stopFridaServer() prefix commands with `su -c`
 *          since frida-server requires root on Android.
 */
class FridaManager(
    private val deviceId: String,
    private val onFindingReady: (JSONObject) -> Unit,
    private val onLog: (String) -> Unit
) {
    private val TAG = "FridaManager"

    private val serverPath = "/data/local/tmp/frida-server"
    private val serverRunning = AtomicBoolean(false)
    private val executor = Executors.newCachedThreadPool()

    // Exposed so MainActivity can route frida_output messages to it
    val scriptRunner = FridaScriptRunner(
        deviceId = deviceId,
        onFindingReady = onFindingReady,
        onLog = onLog
    )
    private val scheduler: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
    private var healthCheckFuture: ScheduledFuture<*>? = null

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    fun start() {
        if (!isFridaServerAvailable()) {
            startFridaServer()
        } else {
            serverRunning.set(true)
            onLog("[FridaManager] frida-server already running")
        }
        // Health check every 30s — auto-restart if crashed
        healthCheckFuture = scheduler.scheduleAtFixedRate({
            val nowRunning = isFridaServerAvailable()
            if (serverRunning.get() && !nowRunning) {
                onLog("[FridaManager] frida-server died, restarting…")
                startFridaServer()
            }
            serverRunning.set(nowRunning)
        }, 30, 30, TimeUnit.SECONDS)
    }

    fun stop() {
        healthCheckFuture?.cancel(false)
        stopFridaServer()
        executor.shutdown()
        onLog("[FridaManager] stopped")
    }

    // ── Command routing ───────────────────────────────────────────────────────

    fun handleCommand(cmd: JSONObject) {
        when (cmd.optString("type")) {
            "frida_inject" -> handleInject(cmd)
            "frida_stop"   -> stopFridaServer()
            "frida_status" -> sendStatus()
            else -> Log.d(TAG, "Unknown frida command: ${cmd.optString("type")}")
        }
    }

    private fun handleInject(cmd: JSONObject) {
        val sessionId = cmd.optString("session_id", "")
        val scriptName = cmd.optString("script_name", "custom")
        val packageName = cmd.optString("package_name", "")

        if (packageName.isEmpty()) {
            onLog("[FridaManager] inject rejected: missing package_name")
            return
        }

        // ARCHITECTURE (Fix #11): The Android app does NOT execute frida commands.
        // frida is a HOST tool. This app:
        //   1. Ensures frida-server is running on the device (so host can connect)
        //   2. Acknowledges the inject command back to the backend
        //
        // The BACKEND (host machine) then runs:
        //   frida -U -f <packageName> -l <scriptFile> --no-pause
        // and streams output back via WebSocket.

        // Ensure frida-server is running so host can connect
        if (!isFridaServerAvailable()) {
            onLog("[FridaManager] frida-server not running, starting for session $sessionId…")
            startFridaServer()
            // Give it 2s to start
            Thread.sleep(2000)
        }

        // Acknowledge to backend — it will now execute frida on the host side
        val ack = JSONObject().apply {
            put("type", "frida_server_ready")
            put("session_id", sessionId)
            put("device_id", deviceId)
            put("package_name", packageName)
            put("script_name", scriptName)
            put("server_running", isFridaServerAvailable())
            put("message", "frida-server ready on device — host can connect via frida -U")
            put("timestamp", System.currentTimeMillis() / 1000L)
        }
        onFindingReady(ack)
        onLog("[FridaManager] Session $sessionId: ready for ${packageName}")
    }

    private fun sendStatus() {
        val running = isFridaServerAvailable()
        serverRunning.set(running)
        val status = JSONObject().apply {
            put("type", "frida_status")
            put("device_id", deviceId)
            put("server_running", running)
            put("server_path", serverPath)
            put("server_binary_exists", File(serverPath).exists())
        }
        onFindingReady(status)
    }

    // ── frida-server process management ───────────────────────────────────────

    private fun startFridaServer() {
        if (!File(serverPath).exists()) {
            onLog("[FridaManager] frida-server binary not found at $serverPath")
            onLog("[FridaManager] Deploy it using: POST /api/mobile/frida/server/push")
            serverRunning.set(false)
            return
        }

        executor.submit {
            try {
                // FIX #10: Use `su -c` — frida-server requires root on Android
                val chmodResult = Runtime.getRuntime()
                    .exec(arrayOf("su", "-c", "chmod +x $serverPath"))
                chmodResult.waitFor(5, TimeUnit.SECONDS)

                // Start frida-server in background mode
                val startProc = Runtime.getRuntime()
                    .exec(arrayOf("su", "-c", "$serverPath -D"))
                serverRunning.set(true)
                onLog("[FridaManager] frida-server started at $serverPath")

                startProc.waitFor()
                serverRunning.set(false)
                onLog("[FridaManager] frida-server exited")

            } catch (e: Exception) {
                serverRunning.set(false)
                onLog("[FridaManager] frida-server start failed: ${e.message}")
                Log.e(TAG, "frida-server start failed", e)
            }
        }
    }

    private fun stopFridaServer() {
        try {
            // FIX #10: Use `su -c` for pkill too
            Runtime.getRuntime()
                .exec(arrayOf("su", "-c", "pkill -f frida-server"))
                .waitFor(5, TimeUnit.SECONDS)
            serverRunning.set(false)
            onLog("[FridaManager] frida-server stopped")
        } catch (e: Exception) {
            Log.w(TAG, "frida-server stop error: ${e.message}")
        }
    }

    /**
     * FIX #9: Use `su -c ps | grep frida-server` to check if the process is
     * running on the Android device. The original code used `frida-ps -U` which
     * is a HOST-MACHINE binary and does not exist in the Android app context.
     */
    fun isFridaServerAvailable(): Boolean {
        // Fast path: if we started it and haven't detected a crash, trust the flag
        if (serverRunning.get()) return true

        // Verify via `ps` (works on rooted devices without su for ps)
        return try {
            val proc = Runtime.getRuntime().exec(arrayOf("su", "-c", "ps -e"))
            val output = proc.inputStream.bufferedReader().readText()
            proc.waitFor(5, TimeUnit.SECONDS)
            output.contains("frida-server")
        } catch (e: Exception) {
            // ps may fail without root; fall back to serverRunning flag
            serverRunning.get()
        }
    }
}

File: ios-companion/Sources/OneInfinityCompanion/Frida/FridaManager.swift
// FridaManager.swift — Frida server management for jailbroken iOS devices
// Phase 4: iOS Support
//
// Architecture (same as Android FridaManager.kt after Fix #11):
//   1. Manages frida-server binary at /usr/bin/frida-server (jailbroken devices)
//   2. Receives {type:frida_inject} commands from backend via WebSocket
//   3. Ensures frida-server is running, sends acknowledgment
//   4. Backend (host) runs `frida -U -f <bundle_id>` against this device
//   5. Receives {type:frida_output} with findings from backend
//
// Requires: jailbroken device with frida-server installed
//           (via Cydia: https://build.frida.re/)

import Foundation

// MARK: - FridaManager

@MainActor
public final class FridaManager: ObservableObject {
    @Published public var serverRunning = false
    @Published public var logs: [String] = []

    private let serverPath = "/usr/sbin/frida-server"
    private let altServerPath = "/usr/bin/frida-server"  // Cydia install location
    private var serverProcess: Process?
    private var healthCheckTimer: Timer?
    private weak var wsManager: WebSocketManager?

    // FridaScriptRunner for processing backend output
    public let scriptRunner = FridaScriptRunner()

    public init(wsManager: WebSocketManager? = nil) {
        self.wsManager = wsManager
    }

    // MARK: Lifecycle

    public func start() {
        if isServerAvailable() {
            serverRunning = true
            appendLog("frida-server already running")
        } else {
            startFridaServer()
        }

        // Health check every 30s
        healthCheckTimer = Timer.scheduledTimer(withTimeInterval: 30.0, repeats: true) { [weak self] _ in
            Task { @MainActor in
                guard let self = self else { return }
                let running = self.isServerAvailable()
                if self.serverRunning && !running {
                    self.appendLog("frida-server died, restarting…")
                    self.startFridaServer()
                }
                self.serverRunning = running
            }
        }
    }

    public func stop() {
        healthCheckTimer?.invalidate()
        stopFridaServer()
        scriptRunner.stopAll()
    }

    // MARK: Command handling

    public func handleCommand(_ json: [String: Any]) {
        let type = json["type"] as? String ?? ""
        switch type {
        case "frida_inject":     handleInject(json)
        case "frida_stop":       stopFridaServer()
        case "frida_status":     sendStatus()
        case "frida_output",
             "frida_session_event",
             "frida_finding_received":
            scriptRunner.processBackendOutput(json, onLog: appendLog(_:))
        default:
            appendLog("Unknown frida command: \(type)")
        }
    }

    private func handleInject(_ json: [String: Any]) {
        let sessionId  = json["session_id"] as? String ?? ""
        let scriptName = json["script_name"] as? String ?? "custom"
        let bundleId   = json["package_name"] as? String ?? ""

        guard !bundleId.isEmpty else {
            appendLog("frida_inject rejected: missing package_name (bundle ID)")
            return
        }

        let needsStart = !isServerAvailable()
        if needsStart {
            appendLog("frida-server not running, starting for session \(sessionId)…")
            startFridaServer()
        }

        // Use async Task.sleep instead of Thread.sleep — never block MainActor.
        // Delay ack by 2s only if we just started the server.
        Task {
            if needsStart {
                try? await Task.sleep(nanoseconds: 2_000_000_000)
            }
            let running = self.isServerAvailable()
            let ack: [String: Any] = [
                "type": "frida_server_ready",
                "session_id": sessionId,
                "device_id": self.deviceIdentifier(),
                "package_name": bundleId,
                "script_name": scriptName,
                "server_running": running,
                "platform": "ios",
                "message": "frida-server ready — host can connect via frida -U -f \(bundleId)",
                "timestamp": Date().timeIntervalSince1970
            ]
            await MainActor.run {
                self.wsManager?.sendFinding(ack)
                self.appendLog("Session \(sessionId): ready for \(bundleId)")
            }
        }
    }

    private func sendStatus() {
        let running = isServerAvailable()
        serverRunning = running
        let status: [String: Any] = [
            "type": "frida_status",
            "device_id": deviceIdentifier(),
            "server_running": running,
            "server_path": resolvedServerPath() ?? "not found",
            "platform": "ios",
            "is_jailbroken": isJailbroken()
        ]
        wsManager?.sendFinding(status)
    }

    // MARK: frida-server process

    private func startFridaServer() {
        guard let path = resolvedServerPath() else {
            appendLog("frida-server not found at \(serverPath) or \(altServerPath)")
            appendLog("Install via Cydia: https://build.frida.re/")
            serverRunning = false
            return
        }

        DispatchQueue.global(qos: .background).async { [weak self] in
            let proc = Process()
            proc.executableURL = URL(fileURLWithPath: path)
            proc.arguments = ["-D"]  // background / daemonize

            do {
                try proc.run()
                self?.serverProcess = proc
                DispatchQueue.main.async {
                    self?.serverRunning = true
                    self?.appendLog("frida-server started at \(path)")
                }
                proc.waitUntilExit()
                DispatchQueue.main.async {
                    self?.serverRunning = false
                    self?.appendLog("frida-server exited")
                }
            } catch {
                DispatchQueue.main.async {
                    self?.serverRunning = false
                    self?.appendLog("frida-server start failed: \(error.localizedDescription)")
                }
            }
        }
    }

    private func stopFridaServer() {
        serverProcess?.terminate()
        serverProcess = nil
        serverRunning = false
        appendLog("frida-server stopped")
    }

    // MARK: Helpers

    public func isServerAvailable() -> Bool {
        if serverRunning { return true }
        // Check via ps
        let ps = Process()
        ps.executableURL = URL(fileURLWithPath: "/bin/ps")
        ps.arguments = ["-e"]
        let pipe = Pipe()
        ps.standardOutput = pipe
        try? ps.run()
        ps.waitUntilExit()
        let output = String(data: pipe.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8) ?? ""
        return output.contains("frida-server")
    }

    private func resolvedServerPath() -> String? {
        if FileManager.default.fileExists(atPath: serverPath) { return serverPath }
        if FileManager.default.fileExists(atPath: altServerPath) { return altServerPath }
        return nil
    }

    public func isJailbroken() -> Bool {
        let jbPaths = [
            "/Applications/Cydia.app",
            "/Library/MobileSubstrate/MobileSubstrate.dylib",
            "/bin/bash",
            "/usr/sbin/sshd",
            "/etc/apt",
            "/private/var/lib/apt"
        ]
        return jbPaths.contains { FileManager.default.fileExists(atPath: $0) }
    }

    private func deviceIdentifier() -> String {
        UIDevice.current.identifierForVendor?.uuidString ?? "ios-unknown"
    }

    private func appendLog(_ msg: String) {
        logs.append("[\(Date().formatted(date: .omitted, time: .standard))] \(msg)")
        if logs.count > 200 { logs.removeFirst() }
        print("[FridaManager] \(msg)")
    }
}

// MARK: - FridaScriptRunner

/// Processes Frida output messages received FROM the backend.
/// Architecture: frida runs on HOST, results come back via WebSocket.
public final class FridaScriptRunner {
    public func processBackendOutput(_ json: [String: Any], onLog: (String) -> Void) {
        let sessionId  = json["session_id"] as? String ?? ""
        let scriptName = json["script_name"] as? String ?? "unknown"
        let output     = json["output"] as? String ?? ""
        let hasFinding = json["has_finding"] as? Bool ?? false

        if !output.isEmpty {
            onLog("[\(scriptName)] \(output.prefix(200))")
        }

        if let finding = json["finding"] as? [String: Any] {
            let vuln     = finding["vulnerability"] as? String ?? "finding"
            let severity = finding["severity"] as? String ?? "?"
            onLog("[\(scriptName)] FINDING [\(severity)]: \(vuln)")
        }

        if json["type"] as? String == "frida_session_event" {
            let event = json["event"] as? String ?? ""
            let count = json["findings_count"] as? Int ?? 0
            onLog("[\(scriptName)] Session \(sessionId) \(event)\(count > 0 ? " — \(count) findings" : "")")
        }
    }

    public func stopAll() {
        // Nothing to stop locally — execution is on the host
    }
}
